### DISCUSSION TITLE
Supporting SignWriting in HarfBuzz

### DISCUSSION THREAD

**Discussion by AmitMY** [harfbuzz/harfbuzz#3360] [General] [answered]
I'm interested in developing a shaper for SignWriting in order to display the SuttonSignWriting font correctly as text, anywhere.

The idea is quite simple, although I'm not sure where to begin, any direction for beginning a new script would be appreciated

### The SignWriting Script
Currently, just showing as a sequence of characters, SignWriting comes in threes:
1. glyph
2. x position
3. y position

**_The rendering of all glyphs is very predictable. Put glyph `glyph` in [`x position`,`y position`]. There are no cases where two glyphs next to each other morph into another, or that under some considerations a glyph position changes._**

For example, here are 12 unicode characters, representing a single word in SignWriting:
![image](https://user-images.githubusercontent.com/5757359/148742074-97d912ac-7341-4658-a70c-794abdd47cab.png)

The first character, `M` is a structural character, telling to create a "box" with center `x` and `y`. There are other box types, but initially I want to support just one, to get this going.

Then, the rest of the symbols can be grouped into 3 groups of 3 characters each:
![image](https://user-images.githubusercontent.com/5757359/148742560-47ffabce-957d-49a7-a711-00fcd93afcce.png)

Telling us where to position each glyph (pseudo code)
![image](https://user-images.githubusercontent.com/5757359/148742642-a3d1589a-dcea-466e-ac4b-e8e06f71e258.png)

Finally we get:
![image](https://user-images.githubusercontent.com/5757359/148742798-e7f534c4-f860-460f-a726-2eefc304fbc8.png)

Here's a design document with more details:
http://www.unicode.org/L2/L2017/17220-signwriting-design-opt.pdf

----

How do I get started at implementing and testing this script in HarfBuzz?

**Comment by behdad** [comment, marked as answer]
Hi there.

We don't believe a new shaper is required. You should be able to use existing GSUB/GPOS mechanisms.

**Comment by AmitMY** [comment]
Thanks @behdad
In that case, I guess I'm not sure where to get started in writing these commands for this font, and what tools (languages/libraries) I use to do so.

It would be really epic if this was supported for sign language, its a long time dream :)

**Comment by behdad** [comment]
I don't design fonts. Others, like @dscorbett or @khaledhosny can advise you on that.

**Comment by dscorbett** [comment]
The positioning is easy. Each character can be shifted up, down, left, or right using contextual single positioning based on the coordinate characters following it.

The hard part will be making the white part of one character hide the character below it, as in your example where the bottom right rim of the face is hidden behind the white part of the fist. There is no general way to do that in OpenType Layout alone. You might have to make a color font.

Note that because this is not the Unicode encoding of SignWriting, you may have some problems relying on the Private Use Area in some applications. It will work in HarfBuzz, though.

**Reply by AmitMY** [reply]
Hmmm I understand the issue with the white area, thanks for pointing that out. 
I do not forsee it to be a real issue though, at least for a proof of concept so we can start tooling for signwriting text in font form, and I guess we can improve otherwise.... 

I hope it is easy :) I will try setting up a Windows VM in order to use VOLT, and I also left another comment below if you want to give it a try ;)

**Comment by khaledhosny** [comment]
If you have a font with the glyph drawn, I might give writing the font layout features a try.

**Reply by AmitMY** [reply]
Thanks @khaledhosny, it would be a dream to get that working...

Here's the ttf font with all the glyphs:
https://github.com/sutton-signwriting/font-ttf/raw/master/src/font/SuttonSignWritingOneD.ttf

And here is the example I gave written in unicode:
`𝠃𝤛𝤵񍉡𝣴𝣵񆄱𝤌𝤆񈠣𝤉𝤚`

There's actually an svg display program we are currently using:
https://slevinski.github.io/SuttonSignWriting/characters/index.html

But it requires a library, and only works on the web

**Reply by Slevinski** [reply]
For One-Dimensional viewing, the Sutton SignWriting One-D font is a great choice.

For Two-Dimensional display, I would suggest using the Sutton SignWriting Line font and the Sutton SignWriting Fill font.  They are available as TTF.  I also have source SVG.  They are written below the baseline to simplify SVG usage.  Characters on plane 4 should be shifted to plane 15 for the line and plane 16 for the fill.

@font-face {
  font-family: "SuttonSignWritingLine";
  src:
    local('SuttonSignWritingLine'),
    url('https://unpkg.com/@sutton-signwriting/font-ttf@1.0.0/font/SuttonSignWritingLine.ttf') format('truetype');
}
@font-face {
  font-family: "SuttonSignWritingFill";
  src:
    local('SuttonSignWritingFill'),
    url('https://unpkg.com/@sutton-signwriting/font-ttf@1.0.0/font/SuttonSignWritingFill.ttf') format('truetype');
}

More details available on www.sutton-signwriting.io if you want them.

**Comment by AmitMY** [comment]
@dscorbett @khaledhosny 
I keep coming back to needing such a font for my university research.
Even on my student salary, I am willing to set a "bounty" of 200$ (PayPal) for a working `ttf` that can only handle `M` boxes and symbol positions, if either of you is willing to try to create one.

**Reply by AmitMY** [reply]
Would there be any reasonable amount of money to make this bounty worthy of an attempt?
